src/plugins/workspace/public/components/workspace_form/select_data_source_panel.tsx

```diff
@@ -3,149 +3,166 @@
+import { DataSourceConnectionTable } from './data_source_connection_table';
 
 export interface SelectDataSourcePanelProps {
   errors?: { [key: number]: WorkspaceFormError };
   savedObjects: SavedObjectsStart;
-  selectedDataSources: DataSource[];
+  assignedDataSources: DataSource[];
   onChange: (value: DataSource[]) => void;
+  showDataSourceManagement: boolean;
 }
 
 export const SelectDataSourcePanel = ({
-  errors,
   onChange,
-  selectedDataSources,
+  assignedDataSources,
   savedObjects,
+  showDataSourceManagement,
 }: SelectDataSourcePanelProps) => {
-  const [dataSourcesOptions, setDataSourcesOptions] = useState<EuiComboBoxOptionOption[]>([]);
-  useEffect(() => {
-    if (!savedObjects) return;
-    getDataSourcesList(savedObjects.client, ['*']).then((result) => {
-      const options = result.map(({ title, id }) => ({
-        label: title,
-        value: id,
+  const [modalVisible, setModalVisible] = useState(false);
+  const [selectedItems, setSelectedItems] = useState<DataSourceConnection[]>([]);
+  const [assignedDataSourceConnections, setAssignedDataSourceConnections] = useState<
+    DataSourceConnection[]
+  >([]);
+  const [toggleIdSelected, setToggleIdSelected] = useState(
+    AssociationDataSourceModalMode.OpenSearchConnections
+  );
+  const {
+    services: { notifications, http, chrome },
+  } = useOpenSearchDashboards<{ CoreStart: CoreStart; workspaceClient: WorkspaceClient }>();
+
+  const handleAssignDataSources = (dataSourceConnections: DataSourceConnection[]) => {
+    setAssignedDataSourceConnections([...assignedDataSourceConnections, ...dataSourceConnections]);
+    setModalVisible(false);
+    const dataSources = dataSourceConnections
+      .filter(
+        ({ connectionType }) => connectionType === DataSourceConnectionType.OpenSearchConnection
+      )
+      .map(({ id, type, name, description }) => ({
+        id,
+        title: name,
+        description,
+        dataSourceEngineType: type as DataSourceEngineType,
       }));
-      setDataSourcesOptions(options);
-    });
-  }, [savedObjects, setDataSourcesOptions]);
-  const handleAddNewOne = useCallback(() => {
-    onChange?.([
-      ...selectedDataSources,
-      {
-        title: '',
-        id: '',
-      },
-    ]);
-  }, [onChange, selectedDataSources]);
+    const savedDataSources: DataSource[] = [...assignedDataSources, ...dataSources];
+    onChange(savedDataSources);
+  };
+
+  const handleUnassignDataSources = (dataSourceConnections: DataSourceConnection[]) => {
+    const savedDataSourcesConnections = (assignedDataSourceConnections ?? [])?.filter(
+      ({ id }: DataSourceConnection) => !dataSourceConnections.some((item) => item.id === id)
+    );
+    setAssignedDataSourceConnections(savedDataSourcesConnections);
+    const savedDataSources = (assignedDataSources ?? [])?.filter(
+      ({ id }: DataSource) => !dataSourceConnections.some((item) => item.id === id)
+    );
+    onChange(savedDataSources);
```

Comment:
May I know in which case the `assignedDataSourceConnections` will not equals to `assignedDataSources`?